src/github.com/couchbase/sync_gateway/rest/admin_api.go

```diff
@@ -39,12 +42,115 @@ func (h *handler) handleCreateDB() error {
 	return base.HTTPErrorf(http.StatusCreated, "created")
 }
 
+
+func (h *handler) handleDbOnline() error {
+	h.assertAdminOnly()
+
+	//If the DB is already trasitioning to: online or is online silently return
+	if (h.db.State == db.DBOnline || h.db.State == db.DBStarting) {
+		return nil
+	}
+
+	//If the DB is currently re-syncing return an error asking the user to retry later
+	if (h.db.State == db.DBReSyncing) {
+		return base.HTTPErrorf(http.StatusServiceUnavailable, "Database _resync is in progress, this may take some time, try again later")
+	}
+
+	body, err := h.readBody()
+	if err != nil {
+		return err
+	}
+
+	var input struct {
+		Delay       int           `json:"delay"`
+	}
+
+	input.Delay = kDefaultDBOnlineDelay;
+
+	json.Unmarshal(body, &input);
+
+	base.LogTo("CRUD", "Taking Database : %v, online in %v seconds", h.db.Name, input.Delay)
+
+	//Set DB state to DBStarting, this wil cause new API requests to be be rejected
+	h.db.State = db.DBStarting
```

Comment:
Won't you need to synchronize changes to h.db.State, either using a mutex or by storing as an atomic uint32?

In general, what's the behaviour if multiple calls to handleDbOnline and handleDbOffline are made at the same time?